File: src/main/java/com/euc/grantfitassessment/eval/pipeline/EvidenceGroundingFilter.java
package com.euc.grantfitassessment.eval.pipeline;

import com.euc.core.EvaluationFilter;
import com.euc.core.EvaluationCriterion;
import com.euc.core.PipelineContext;
import com.euc.grantfitassessment.pipeline.GrantFitContextKeys;

import java.util.ArrayList;
import java.util.List;

/**
 * Scores EUC evaluation stage "evidenceGrounding": every expected evidence
 * keyword (established independently as ground truth) must appear in the
 * reasoner's cited supportingEvidence. Same vacuous-pass rule as
 * ProgramAlignmentFilter for ineligible cases — nothing was reasoned about,
 * so there is nothing to check evidence for.
 */
public class EvidenceGroundingFilter implements EvaluationFilter {

    @Override
    @SuppressWarnings("unchecked")
    public Verdict evaluate(PipelineContext context, EvaluationCriterion stage) {
        boolean actualEligible = context.get(GrantFitContextKeys.ELIGIBLE, Boolean.class);
        if (!actualEligible) {
            context.put(GrantFitEvalContextKeys.MISSING_EVIDENCE_KEYWORDS, List.of());
            return Verdict.PASSED;
        }

        List<String> evidence = context.get(GrantFitContextKeys.SUPPORTING_EVIDENCE, List.class);
        List<String> expectedKeywords = context.get(GrantFitEvalContextKeys.EXPECTED_EVIDENCE_KEYWORDS, List.class);

        List<String> missing = new ArrayList<>();
        for (String keyword : expectedKeywords) {
            boolean found = evidence.stream().anyMatch(e -> e.toLowerCase().contains(keyword.toLowerCase()));
            if (!found) {
                missing.add(keyword);
            }
        }

        context.put(GrantFitEvalContextKeys.MISSING_EVIDENCE_KEYWORDS, missing);
        return missing.isEmpty() ? Verdict.PASSED : Verdict.FAILED;
    }
}

File: src/euc/grantfitassessment/eval/pipeline/evidence_grounding_filter.py
"""EVAL-EVIDENCE — every expected keyword must appear as a case-insensitive
substring of at least one supporting-evidence entry. Same vacuous-pass-on-
ineligible rule as EVAL-ALIGNMENT."""

from __future__ import annotations

from euc.core.context import PipelineContext
from euc.core.filters import Verdict
from euc.core.models import EvaluationCriterion
from euc.grantfitassessment import context_keys as prod_keys
from euc.grantfitassessment.eval import context_keys as eval_keys


def evidence_grounding_filter(context: PipelineContext, criterion: EvaluationCriterion) -> Verdict:
    actual_eligible: bool = context.get(prod_keys.ELIGIBLE)
    if not actual_eligible:
        context.put(eval_keys.MISSING_EVIDENCE_KEYWORDS, [])
        return Verdict.PASSED

    evidence: list[str] = context.get(prod_keys.SUPPORTING_EVIDENCE)
    expected_keywords: list[str] = context.get(eval_keys.EXPECTED_EVIDENCE_KEYWORDS)

    missing = [
        keyword
        for keyword in expected_keywords
        if not any(keyword.lower() in e.lower() for e in evidence)
    ]

    context.put(eval_keys.MISSING_EVIDENCE_KEYWORDS, missing)
    return Verdict.PASSED if not missing else Verdict.FAILED
